### DISCUSSION TITLE
使用Kubernetes部署报socket read timeout exception，请问有可能是什么问题？？

### DISCUSSION THREAD

**Discussion by isaacV1995** [apolloconfig/apollo#3778] [Q&A]
错误信息：
2021-06-23 14:06:15.230 ERROR 1 --- [a-server-dev-11] c.n.e.cluster.ReplicationTaskProcessor   : It seems to be a socket read timeout exception, it will retry later. if it continues to happen and some eureka node occupied all the cpu time, you should set property'eureka.server.peer-node-read-timeout-ms' to a bigger value

不是用的helm，是自己用代码打包的镜像，然后按照/scripts/apollo-on-kubernetes/README.md部署的

**Comment by nobodyiam** [comment]
可以看看 config service cpu, mem, gc 等情况

**Reply by isaacV1995** [reply]
![image](https://user-images.githubusercontent.com/38184495/123186874-a8c71100-d4cb-11eb-8536-7033faac1879.png)
![image](https://user-images.githubusercontent.com/38184495/123186879-aa90d480-d4cb-11eb-9783-1b67b7850d6e.png)
![image](https://user-images.githubusercontent.com/38184495/123186883-abc20180-d4cb-11eb-9b00-01c195f3d755.png)
这是部署config service的cpu和mem情况，从占用率看的话应该没什么问题。config service的界面显示是这样的：
![image](https://user-images.githubusercontent.com/38184495/123187127-1d01b480-d4cc-11eb-8d63-af152cf23606.png)

**Reply by isaacV1995** [reply]
修改eureka.server.peer-node-read-timeout-ms参数到300ms，然后不报错了，具体问题还得找一下，单机部署没有报这个错

**Reply by nobodyiam** [reply]
可以等到容器里面互相 curl 一下看看响应时间

**Reply by isaacV1995** [reply]
好的，进容器看了一下日志，config server的master节点和两个子节点网络不通，导致连接超时了，把这个弄好再看看

**Reply by isaacV1995** [reply]
容器间的网络弄通了，portal的服务也起来了，能使用，但是还有个问题：部署了两个环境的Kubernetes，有一个环境的config server还是会报out of serivece，去看这个异常的config server的日志也没报错
有问题的环境：
![image](https://user-images.githubusercontent.com/38184495/123568846-c233ca00-d7f7-11eb-9b30-87cd96b0d4de.png)
正常的环境：
![image](https://user-images.githubusercontent.com/38184495/123568892-dbd51180-d7f7-11eb-8ca7-bc0b79287aa7.png)
有问题的config server curl其他两个config server的结果：
![image](https://user-images.githubusercontent.com/38184495/123570096-235c9d00-d7fa-11eb-931e-21b52b8ef80d.png)

**Reply by nobodyiam** [reply]
看看另外两个 config service 的日志中是否有错误？

**Reply by nobodyiam** [reply]
另外好奇问下，helm chart 使用了 k8s 的原生服务发现，所以 config service 不需要部署为 stateful set，降低了部署难度，你选择使用 stateful set 部署而不是 helm chart 部署有啥特殊考虑吗？

**Reply by isaacV1995** [reply]
部署是让运维的同事帮忙部署的，因为改动了代码，就自己打包了镜像，然后在scripts里面找到了apollo-on-kubernetes的readme发给了他，他说这个是和helm chart部署效果是一样。

所有的config service的日志都没报错，然后我看另一个环境是正常的，就让他重新走了一遍部署流程看看，结果这次变成了一个up，两个out_of_service了。。。准备弄完手里的活，学一下Kubernetes看看怎么部署了orz

**Reply by isaacV1995** [reply]
之前他也试过用helm chart部署了官方发布的apollo，那个是没问题的

**Reply by nobodyiam** [reply]
自己打包的镜像也可以使用 helm chart 部署的，指定 image 就可以了

**Reply by isaacV1995** [reply]
好的，我试试

**Reply by isaacV1995** [reply]
我用helm chart部署好了，然后运维的同事和我说生产环境不能连外网，不能装helm，人傻了。。。感觉只能先这么用着了，portal功能是正常的，就是config server有两个不是up状态，日志里面也没报错

**Reply by isaacV1995** [reply]
今天又弄了一会部署为stateful set的情况，发现部署的config server都是至少能up一个节点，但是又不是所有节点都up。然后对比了一下stateful set和helm chart的配置文件，发现data.application-github.properties有区别。
stateful set里面有配置eureka.service.url：
![image](https://user-images.githubusercontent.com/38184495/124079534-dbe54380-da7b-11eb-9cfc-d6c8f71d3e2f.png)
helm chart里面没有这个参数的配置，多了apollo.config-service.url和apollo.admin-service.url：
![image](https://user-images.githubusercontent.com/38184495/124079637-fc150280-da7b-11eb-9b76-bda0714c5195.png)

请问下这个会有影响吗？

**Reply by nobodyiam** [reply]
这两个配置是有区别的，helm chart 是用的 k8s 原生的服务发现，所以不用配置 eureka.service.url 了。

**Reply by isaacV1995** [reply]
感觉是eureka的问题，现在日志里面只能看到这样的消息，请问下apollo能输出debug级别的日志吗，部署文档里好像没看到修改输出日志等级的配置。
![image](https://user-images.githubusercontent.com/38184495/124254652-dc550b80-db5b-11eb-90f0-88570e46e513.png)

**Reply by nobodyiam** [reply]
apollo 是一个普通的 spring boot 程序，调整 Logging level 可以参考[spring boot 文档](https://docs.spring.io/spring-boot/docs/current/reference/html/features.html#features.logging)

**Reply by isaacV1995** [reply]
好的~
